_posts/2025-04-23-#31444.md

```diff
@@ -10,17 +10,95 @@ status: upcoming
+[PR #31444](https://github.com/bitcoin/bitcoin/pull/31444) adds `TxGraph` functionality to assist in selecting
+transactions for block building and eviction:
+    - `TxGraph` adds a chunk index, a total ordering of all chunks in the graph across all clusters. A
+      [`ChunkOrder`
+comparator](https://github.com/bitcoin-core-review-club/bitcoin/commit/3429e9d79df1336cf1d0a61cb5f9bf028aa860b2#diff-c25ebcc6e4dc9c9bb21462f796fbb470cee3f0ff1e77bfc4033f4e8460f4482cR282)
+defines the order in which chunks would be mined.
+    - `BlockBuilder` can just iterate through the chunks in order. Also, since a total orderering of all graph entries
+      exists, we can get the “least likely to be mined” transaction without building a full mempool worth of block
+templates: simply take the last chunk in the index.
+    - `TxGraph` exposes a `BlockBuilder` to iterate over the chunks from best to worst. `BlockBuilder` can `Skip` chunks (i.e.
+      if they are too large to fit in the remaining block space), and those chunks' descendants will be subsequently
+skipped as well.
+    - `TxGraph` also exposes a `GetWorstMainChunk` method to identify the transactions that would be included in a block last.
+- The PR also adds a `GetMainStagingDiagrams` method to make it possible to compare the feerate diagram of the `TxGraph`
+  with and without its staged changes.
+- We have reviewed `TxGraph` basic functionality in a previous [meeting](/31363). If `TxGraph` is new to you, it may be
+  easier to get feel for `TxGraph` by first reviewing `SimTxGraph` in the fuzzer, which has similar functionality but a
+far simpler design:
+    - It uses a single `DepGraph` to represent all transactions across all clusters, which means it doesn’t need to implement merges or track much information about the set of clusters.
+    - It implements the staging level by creating a copy of main and deleting either the main or the staing to commit or abort staged changes. This approach is more memory-intensive but doesn’t need to track the differences between the levels.
 
 ## Questions
 
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+1. Why are block building and eviction relevant to each other? Wouldn’t it be easier to evict transactions by the order
+   they entered the mempool?
+1. What is `GetMainStagingDiagrams` useful for? Why might we exclude the clusters that are identical in main and staging?
+1. True / false: if all clusters are singletons (have 1 transaction each), `m_main_chunkindex` would just be sorting
+   the transactions in order by feerate.
+1. In english, using the approach in this PR, what is the algorithm for selecting transactions in order for block building? And for eviction?
+1. In pseudocode, how would a client of `BlockBuilder` use it to build a block? When would `GetCurrentChunk`, `Include`, and `Skip` be called?
```

Comment:
I like the idea, but unless it's a simple one-liner, this might get messy if everyone starts dumping lines of code?

Reply:
Ok, removed the "in pseudocode" part. Expecting a freeform thing of 1-2 sentences.